### ISSUE TITLE
Improve runtime for small areas

### ISSUE THREAD

**Issue by msbarry** [onthegomap/planetiler#478]
Planetiler takes ~30 seconds to run even for the smallest areas (like andorra from geofabrik). Let's see if there is any way to improve that.  Here's a summary of runtime over andorra:

```
0:00:33 INF -   overall          33s cpu:1m12s gc:3s avg:2.2
0:00:33 INF -   lake_centerlines 3s cpu:12s gc:1s avg:4.4
0:00:33 INF -     read     1x(35% 0.9s done:2s)
0:00:33 INF -     process  9x(1% 0s wait:1s done:2s)
0:00:33 INF -     write    1x(0% 0s wait:1s done:2s)
0:00:33 INF -   water_polygons   12s cpu:17s avg:1.4
0:00:33 INF -     read     1x(94% 12s)
0:00:33 INF -     process  9x(0% 0s wait:12s)
0:00:33 INF -     write    1x(0% 0s wait:12s)
0:00:33 INF -   natural_earth    11s cpu:14s avg:1.3
0:00:33 INF -     read     1x(66% 7s done:4s)
0:00:33 INF -     process  9x(2% 0.2s wait:7s done:4s)
0:00:33 INF -     write    1x(0% 0s wait:8s done:4s)
0:00:33 INF -   osm_pass1        0.4s cpu:2s avg:3.7
0:00:33 INF -   osm_pass2        1s cpu:5s avg:5
0:00:33 INF -     read     1x(0% 0s)
0:00:33 INF -     process  9x(31% 0.3s)
0:00:33 INF -     write    1x(3% 0s wait:1s)
0:00:33 INF -   boundaries       0s cpu:0.1s avg:2.9
0:00:33 INF -   sort             0.1s cpu:0.7s avg:7.2
0:00:33 INF -   archive          0.5s cpu:3s avg:5.6
```

**Comment by msbarry** [comment]
The biggest issues are natural earth and water polygons since planetiler has to deserialize every feature for the whole planet.  One idea would be to switch natural earth to read the geopackage source, and use the built-in spatial index to limit what we read to only what's inside the bounding box.  I'm not sure if we could do something similar with water polygons though?

cc/ @bdon

**Comment by bdon** [comment]
At the most extreme we could define a `ReadableTileArchive` as another input type that is passed directly as tiled features, without touching the FeatureCollector API; the OSM or NE-derived ocean is going to be exactly the same for every planetiler output modulo tags/buffer sizes. That would make water polygons cost effectively nothing. 

Otherwise we might be able to read the Shapefile index if one is included for water polygons, or migrate to another indexed format for it (Geopackage, FGB?)

**Comment by msbarry** [comment]
Another hybrid option might be to compute a spatial index the first time we read a file and use it to speed up subsequent reads?

**Comment by msbarry** [comment]
Or ask the maintainers of the water polygons source if they'd be up for distributing in geopackage format/adding a spatial index.

**Comment by msbarry** [comment]
Another low hanging fruit would be to keep the unzipped file contents around between runs

**Comment by bdon** [comment]
Context on Geopackage etc output from osmcoastline: https://github.com/osmcode/osmcoastline/issues/35#issuecomment-660488805

**Comment by erik** [comment]
For the Natural Earth / geopackage case, we can also have profiles declare the limited set of tables that they're interested in and skip reading features that won't be processed.

**Comment by bdon** [comment]
Proposal here https://github.com/onthegomap/planetiler/pull/635

This addresses Natural Earth and other bring-your-own geopackages. 

For water and land polygons I'd be happy to mirror those via cloud storage bucket in indexed GPKG format, or we can attempt to add that to the upstream data source.

**Comment by Nnnes** [comment]
https://github.com/onthegomap/planetiler/blob/2f1573ec7c5681a8ebc51ddb5e01db27056e6107/planetiler-core/src/main/java/com/onthegomap/planetiler/reader/SourceFeatureProcessor.java#L76-L86

The `getFeatureCount()` call on line 78 here can increase runtime for small areas significantly when Planetiler is run with default settings (i.e. no args except `--area=...`). In particular, at one point it unzips natural_earth_vector.sqlite.zip (currently 434MB zipped/849MB unzipped) to count features, but then the unzipped file is ignored by the `pipeline` a few lines later, which unzips the same file into the same location a second time. This is visible in the console output.

Replacing line 78 with `long featureCount = 0;` reduces runtime by ~15 seconds on my machine (41s -> 26s for Andorra), or ~6 seconds on first run with `--natural-earth-keep-unzipped` (32s -> 26s).

The only place `featureCount` is used is in `addRatePercentCounter("read", ...)`. Perhaps this could just be `addRateCounter(...)`?

**Comment by msbarry** [comment]
Yeah unzipping natural earth twice is kind of annoying... a couple of other options:
- Use shapefile natural earth distribution that can be streamed directly from zip file without unpacking first
- have the natural earth download step unzip the SQLite file so it's only done once
- use geopackage distribution to let us use the spatial index to limit features read (it's still a SQLite file that needs to be unzipped though)

Even if we skip showing % complete we still need to unzip it to read features. 

OSM water polygons are similar size but we use shapefile distribution to avoid the unzipping cost. @bdon mentioned eventually using geopackage for that too so we can use spatial index.  Overture water source is another possibility.

**Comment by flymarq** [comment]
Why is there no option to disable these external data sources? Some of them are quite outdated ...

I tried different cmdline options, but without these data planetiler refuses to start.
